# LongRangeHaplotype Estimates (GenomeSelectionScan)


Phasing with Whatshap + Shapeit4

```bash

cp ../VCF_Filter/NoRelate/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz .

cp ../Fst/samples .

cat samples | while read line ;
do echo ${line} ;
sbatch Whatshap.sh ${line} ;
done

```

Whatshap.sh

```bash

mkdir temp.${1}
cp SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz temp.${1}/
cd temp.${1}

module purge
module load BCFtools/1.9-intel-2018b

bcftools view -s ${1} -O z -o ${1}_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz

module purge
module load WhatsHap/0.18-foss-2019a-Python-3.7.2

whatshap phase -o ${1}_NoRelate.semi-phased.vcf.gz --tag=PS --reference=Puffin_ReferenceGenome.NU.MT.fasta ${1}_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
${1}.Puffin_ReferenceGenome.NU.MT.realigned.bam 

cp ${1}_NoRelate.semi-phased.vcf.gz  ../
cd ../
rm -r temp.${1}

```

```bash

#merge semi-phased
module purge
module load BCFtools/1.9-intel-2018b

cat samples | while read line ;
do echo ${line} ;
gunzip ${line}_NoRelate.semi-phased.vcf.gz ;
bgzip ${line}_NoRelate.semi-phased.vcf ;
bcftools index ${line}_NoRelate.semi-phased.vcf.gz ;
done

bcftools merge -O z -o SPI_BJO_ROS_semi-phased.merged.vcf.gz \
BJO001_NoRelate.semi-phased.vcf.gz \
BJO003_NoRelate.semi-phased.vcf.gz \
BJO005_NoRelate.semi-phased.vcf.gz \
BJO006_NoRelate.semi-phased.vcf.gz \
BJO007_NoRelate.semi-phased.vcf.gz \
BJO008_NoRelate.semi-phased.vcf.gz \
ROS001_NoRelate.semi-phased.vcf.gz \
ROS002_NoRelate.semi-phased.vcf.gz \
ROS003_NoRelate.semi-phased.vcf.gz \
ROS006_NoRelate.semi-phased.vcf.gz \
ROS007_NoRelate.semi-phased.vcf.gz \
ROS008_NoRelate.semi-phased.vcf.gz \
SPI001_NoRelate.semi-phased.vcf.gz \
SPI003_NoRelate.semi-phased.vcf.gz \
SPI004_NoRelate.semi-phased.vcf.gz \
SPI011_NoRelate.semi-phased.vcf.gz \
SPI015_NoRelate.semi-phased.vcf.gz

bcftools index SPI_BJO_ROS_semi-phased.merged.vcf.gz

#Shapeit4 Phasing

for i in {01..24} ;
do echo ${i} ;
sbatch Shapeit4.sh ${i} ;
done

```

Shapeit4.sh

```bash

module purge
module load SHAPEIT4/4.1-gompi-2019a

shapeit4 --input SPI_BJO_ROS_semi-phased.merged.vcf.gz \
--region Chr_${1} \
--use-PS 0.0001 \
--thread 4 \
--log Chr_${1}.phased.log \
--sequencing \
--output SPI_BJO_ROS.NoRelate.Chr_${1}.phased.vcf.gz

```

```bash

module purge
module load BCFtools/1.9-intel-2018b

ls SPI_BJO_ROS.NoRelate.Chr_*.phased.vcf.gz > file.list

for i in {01..24} ;
do echo ${i} ;
bcftools index SPI_BJO_ROS.NoRelate.Chr_${i}.phased.vcf.gz ;
done

bcftools concat -n -f file.list -Oz -o SPI_BJO_ROS.NoRelate.Autosomes.phased.vcf.gz
bcftools index SPI_BJO_ROS.NoRelate.Autosomes.phased.vcf.gz

grep "BJO" samples > bjo
grep "SPI" samples > spi
grep "ROS" samples > ros

bcftools view -S bjo -O z -o BJO.Autosomes.phased.vcf.gz SPI_BJO_ROS.NoRelate.Autosomes.phased.vcf.gz
bcftools index BJO.Autosomes.phased.vcf.gz
bcftools view -S spi -O z -o SPI.Autosomes.phased.vcf.gz SPI_BJO_ROS.NoRelate.Autosomes.phased.vcf.gz
bcftools index SPI.Autosomes.phased.vcf.gz
bcftools view -S ros -O z -o ROS.Autosomes.phased.vcf.gz SPI_BJO_ROS.NoRelate.Autosomes.phased.vcf.gz
bcftools index ROS.Autosomes.phased.vcf.gz

mkdir R
cd R
cp ../ROS.Autosomes.phased.vcf.gz .
cp ../SPI.Autosomes.phased.vcf.gz .
cp ../BJO.Autosomes.phased.vcf.gz .
```

R on HPC

Rehh.R

```bash

#!/usr/bin/env Rscript

.libPaths(c("/cluster/projects/nn9244k/for_oliver/R",.libPaths()))
library("unixtools")
library("rehh")
library("tidyverse")

args <- commandArgs(trailingOnly=TRUE)
pop <- args[1]
i <- args[2]
set.tempdir(paste("/cluster/work/users/oliverke/HighCov/NucAnalysis/GATK/LRH_estimates/R/temp.",pop,"_",i,sep = ""))

hh <- data2haplohh(hap_file = paste(pop,".Autosomes.phased.vcf.gz", sep = ""),
                   polarize_vcf = FALSE, 
                   vcf_reader = "data.table" ,
                   chr.name = paste("Chr_",i,sep = ""))
# perform scan on a single chromosome (calculate iHH values)
hh_f <- subset(hh, min_maf = 0.05)
rm(hh)
scan <- scan_hh(hh_f, polarized = FALSE)

write.table(scan, file = paste(pop,".scan_Chr_",i,".txt",sep = ""), quote = FALSE, row.names = FALSE)
```

Continue

```bash

for pop in SPI ROS BJO ;
do echo ${pop} ;
for i in {01..24} ;
do echo ${i}; 
sbatch Rehh.sh ${pop} ${i} ;
done ;
done ;
```

Rehh.sh

```bash


module load R/4.1.0-foss-2021a

mkdir /cluster/work/users/oliverke/HighCov/NucAnalysis/GATK/LRH_estimates/R/temp.${1}_${2}

Rscript Rehh.R ${1} ${2}

```

Then

```bash

rm slurm*
rm -r temp*

for pop in SPI BJO ROS ;
do echo ${pop} ;
head -1 ${pop}.scan_Chr_01.txt > ${pop}.scan_All.txt ;
for i in {01..24} ; 
do echo ${i} ;
tail -n+2 ${pop}.scan_Chr_${i}.txt >> ${pop}.scan_All.txt ;
done ;
done ;

#LOCAL
#do Rehh1.R


#bedtools for sliding window
awk -v OFS='\t' {'print $1,$2'} Puffin_ReferenceGenome.NU.MT.fasta.fai > test.txt
cat test.txt | grep -v "Chr_W" | grep -v "Chr_Z" | grep -v "Chr_Unplaced" | grep -v "Chr_MT" | grep -v "Chr_25" > Puffin_genomeFile.txt
rm test.txt

module purge
module load BEDTools/2.30.0-GCC-10.2.0

bedtools makewindows -g Puffin_genomeFile.txt -w 50000 -s 25000 > windows.bed

awk -F ' ' '{print $1"\t"$8"\t"$2"\t"$7}' SPI_IHS_bedlike.tsv | tail -n+2 | grep -v "NA" > SPI.IHS.bed
bedtools intersect -a windows.bed -b SPI.IHS.bed -wa -wb > SPI.IHS.WindowsToScores.bed
bedtools groupby -i SPI.IHS.WindowsToScores.bed -g 1,2,3 -c 7 -o mean > SPI.IHS.AveScoresPerWindow.bed

awk -F ' ' '{print $1"\t"$10"\t"$2"\t"$9}' SPI_ROS_xpEHH_bedlike.tsv | tail -n+2 | grep -v "NA" > SPI.ROS.xpEHH.bed
bedtools intersect -a windows.bed -b SPI.ROS.xpEHH.bed -wa -wb > SPI.ROS.xpEHH.WindowsToScores.bed
bedtools groupby -i SPI.ROS.xpEHH.WindowsToScores.bed -g 1,2,3 -c 7 -o mean > SPI.ROS.xpEHH.AveScoresPerWindow.bed

#Rehh2.R on local



```