# STRs

with HIPSTR 

```bash


mkdir STR
cd STR

#Build HipSTR reference
mkdir TRF
cd TRF
mkdir raw_fasta trf_results fixed_trf_results
cd raw_fasta
cp ../../../Ref/Puffin_ReferenceGenome.NU.MT.FINAL.fasta .
#split up into chromosome fastas
module load seqtk/1.3-foss-2018b
for chrom in {01..24} Z W Unplaced ;
do echo ${chrom} ;
echo -e "Chr_${chrom}" > name.lst ;
seqtk subseq Puffin_ReferenceGenome.NU.MT.FINAL.fasta name.lst > chr${chrom}.fa ;
rm name.lst ;
done

cd ../../

for chrom in {01..24} Z W Unplaced ;
do echo ${chrom} ;
sbatch run_TRF.sh TRF/raw_fasta/chr${chrom}.fa TRF/trf_results 5 ;
done

```

nano run_TRF.sh

```bash


# Input parameters
INPUT_FILE=$1
OUTPUT_DIR=$2
MIN_SCORE=$3

# Default parameters to use with Tandem Repeat Finder
MATCH_WT=2
MISMATCH_PEN=7
INDEL_PEN=7
P_MATCH=80
P_INDEL=10
MAX_PERIOD=500

NAME=$(basename $INPUT_FILE)

module load TRF/4.09-linux64

trf $INPUT_FILE $MATCH_WT $MISMATCH_PEN $INDEL_PEN $P_MATCH $P_INDEL $MIN_SCORE $MAX_PERIOD -h -d -l 6 -ngs > $OUTPUT_DIR/$NAME
```

THEN

```bash

rm slurm*

module purge
module load HipSTR/0.6.2-foss-2018b-Python-3.6.6

for chrom in {01..24} Z W Unplaced ;
do echo ${chrom} ;
python2 fix_trf_output.py TRF/trf_results/chr${chrom}.fa TRF/fixed_trf_results/chr$chrom.fa ;
done ;

```

nano fix_trf_output.py

```python

import sys

# Indices for various fields used in pattern correction
pattern_index   = 13
period_index    = 2
nrepeat_index   = 3
cons_size_index = 4

# Indices for fields used in overlap resolution 
start_index = 0
stop_index  = 1
score_index = 7

# Number of tokens in a correct input line
num_tokens = 17

# Run the Z-algorithm on the provided sequence. Returns an array of length len(NMER), where arr[i] corresponds
# to the length of the sequence starting at position i that is a prefix of NMER
def zalgorithm(nmer):
    res = len(nmer)*[0]
    l   = -1
    r   = -1

    for i in range(1, len(nmer), 1):
        if i > r:
            j = i
            while j < len(nmer) and nmer[j] == nmer[j-i]:
                j = j + 1

            res[i] = j-i
            l      = i
            r      = j-1
        else:
            i_prime = i-l
            beta    = r-i+1

            if res[i_prime] < beta:
                res[i] = res[i_prime]
            elif res[i_prime] == beta:
                j = r + 1
                while j < len(nmer) and nmer[j] == nmer[j-i]:
                    j = j + 1

                res[i] = j-i
                l      = i
                r      = j-1                
            else:
                res[i] = beta
    res[0] = len(nmer)
    return res

# Returns the repeating subunit of the provided sequence if it consists of exactly 2 or more copies. Otherwise, returns false
def has_subseq(nmer):
    prefix_lengths = zalgorithm(nmer)

    for k in range(1, len(nmer), 1):
        if len(nmer) % k == 0:
            match = True

            for segment in range(len(nmer)/k):
                coord = segment*k
               
                if prefix_lengths[coord] < k:
                    match = False
                    break

            if match:
                return nmer[0:k]
    return False

# Create a new file containing the same information as contained in INPUT_FILE,  except that the period, pattern and number of repeats for entries 
# with incorrect patterns are corrected. Also filters out entries with a corrected period greater than MAX_PERIOD
def correct_pattern_errors(input_file, output_file, max_period=6):
    data    = open(input_file, "r")
    output  = open(output_file, "w")

    # Read the chromosome heading and write it out to the filtered file
    output.write(data.readline())
 
    fail_count    = 0
    skip_count    = 0
    correct_count = 0

    for line in data:
        tokens = line.strip().split()
        
        if len(tokens) != num_tokens:
            exit("ERROR: Malformed input file")
            skip_count = skip_count + 1
            continue

        pattern = tokens[pattern_index]
        subseq  = has_subseq(pattern)
        if subseq:
            new_patt = subseq
            tokens[nrepeat_index]   = str(float(tokens[nrepeat_index])*len(pattern)/len(new_patt))
            tokens[pattern_index]   = new_patt
            tokens[period_index]    = str(len(new_patt))
            tokens[cons_size_index] = str(len(new_patt))
            new_line = ' '.join(tokens) + '\n'
            
            if len(new_patt) <= max_period:
                correct_count += 1
                output.write(new_line)
            else:
                fail_count += 1
        else:
            if int(tokens[cons_size_index]) <= max_period:
                output.write(line)
            else:
                fail_count += 1

    data.close()
    output.close()
    print("Modified %d records whose patterns were incorrect"%(correct_count))
    print("Removed  %d records whose patterns had a PERIOD > %d"%(fail_count, max_period))

def main():
    if len(sys.argv) != 3:
        exit("ERROR: This program accepts exactly two arguments: the input and output file names. Exiting...")
    
    print("FIXING FILE " + sys.argv[1])
    correct_pattern_errors(sys.argv[1], sys.argv[2], max_period=6)

if __name__ == "__main__":
    main()
```

Then

```bash

module purge
module load HipSTR/0.6.2-foss-2018b-Python-3.6.6

files=""
for chrom in {01..24} Z W Unplaced ;
do echo ${chrom} ;
files="${files},TRF/fixed_trf_results/chr${chrom}.fa" ;
done ;
files=`echo ${files} | sed "s/,//"` 
python2 trf_parser.py ${files} > filtered_repeats.puffin.bed

module purge
module load BEDTools/2.27.1-foss-2018b
bedtools sort -i filtered_repeats.puffin.bed > filtered_repeats.puffin.sorted.bed

module purge
module load HipSTR/0.6.2-foss-2018b-Python-3.6.6
python2 analyze_overlaps.py filtered_repeats.puffin.sorted.bed pass.puffin fail.puffin
```

nano trf_parser.py

```python

import sys

num_tokens     = 17
score_index    = 7
period_index   = 4  # Actually the size of the consensus sequence, but this is more accurate
motif_index    = 13
sequence_index = 14
start_index    = 0
stop_index     = 1
nrepeat_index  = 3

def min_perm(seq):
    min_perm = seq
    for i in xrange(len(seq)):
        other = seq[i:]+seq[0:i]
        if other < min_perm:
            min_perm = other
    return min_perm

def rev_complement(seq):
    rev     = seq[::-1]
    mapping = {"A":"T", "T":"A", "C":"G", "G":"C", "N":"N"}
    res     = ""
    for i in xrange(len(rev)):
        res = res+mapping[rev[i]]
    return res

def canonical_motif(motif):
    return min(min_perm(motif), min_perm(rev_complement(motif)))

def create_filtered_trf_bed_file(input_files):
    max_period        = 6
    period_vals       = [1,  2,  3,  4,  5,  6]
    period_thresholds = [20, 22, 28, 28, 32, 34]
    filt_count        = 0
    keep_count        = 0

    for i in range(len(input_files)):
        data  = open(input_files[i], "r")
        chrom = data.readline().strip()[1:]
        for line in data:
            tokens = line.strip().split()
            period = int(tokens[period_index])
            if period > max_period:
                filt_count += 1
                continue

            score = int(tokens[score_index])
            if score <= period_thresholds[period-1]:
                filt_count += 1
                continue

            new_tokens    = [chrom] + list(map(lambda x: tokens[x], [start_index, stop_index, period_index, motif_index, nrepeat_index, score_index, sequence_index]))
            new_tokens[4] = min_perm(new_tokens[4])
            print("\t".join(new_tokens))
            keep_count += 1

        data.close()
    sys.stderr.write("Kept %d out of %d records (%.2f%%) with sufficiently high scores\n"%(keep_count, keep_count+filt_count, 100.0*keep_count/(keep_count+filt_count)))
    return

def main():
    if len(sys.argv) != 2:
        exit("ERROR: Program requires exactly one arguments: a comma-separated list of file paths")
    files  = sys.argv[1].strip().split(",")
    create_filtered_trf_bed_file(files)

if __name__ == "__main__":
    main()
```

nano analyze_overlaps.py

```python

import sys

def main():
    data       = open(sys.argv[1], "r")
    merge_pass = open(sys.argv[2], "w")
    merge_fail = open(sys.argv[3], "w")
    pass_count = 0
    fail_count = 0
    regions    = []
    cur_chrom  = ""
    min_start  = 0
    max_stop   = -1        

    for line in data:
        tokens              = line.strip().split("\t")
        chrom               = tokens[0]
        start, stop, period = map(int, tokens[1:4])
        motif               = tokens[4]
        num_repeats         = float(tokens[5])
        score               = float(tokens[6])
        sequence            = tokens[7]

        if chrom != cur_chrom or start > max_stop:
            # Analyze the previous set of overlapping regions
            region_size = max_stop - min_start + 1
            max_score   = 0
            max_index   = -1
            cov_frac    = 0.0
            for index,region in enumerate(regions):
                if region[5] >= max_score:
                    max_index = index
                    max_score = region[5]
                    cov_frac  = 1.0*(region[2]-region[1]+1)/region_size

            if max_index != -1:
                region = regions[max_index]
                if cov_frac > 0.85:
                    merge_pass.write("%s\t%s\t%d\t%d\t%.1f\t%s\n"%(region[0], region[1], region[2], region[3], region[4], region[6]))
                    pass_count += (len(regions) > 1)
                elif len(set(list(map(lambda x: x[3], regions)))) == 1:
                    all_motifs   = [reg[6] for reg in regions]
                    merge_pass.write("%s\t%s\t%d\t%d\t%.1f\t%s\n"%(region[0], min_start, max_stop, region[3], 1.0*(max_stop-min_start+1)/int(region[3]), "/".join(all_motifs)))
                else:
                    for region in regions:
                        merge_fail.write("%s\t%s\t%d\t%d\t%.1f\t%d\t%s\t%s\n"%(region[0], region[1], region[2], region[3], region[4], region[5], region[6], region[7]))
                    fail_count += (len(regions) > 1)

            # Reset the region info
            regions   = []
            min_start = start
            max_stop  = stop
            cur_chrom = chrom
        else:
            max_stop = max(max_stop, stop)

        regions.append((chrom, start, stop, period, num_repeats, score, motif, sequence))

    data.close()
    merge_pass.close()
    merge_fail.close()

if __name__ == "__main__":
    main()
```

Then

```bash
module purge
module load BEDTools/2.27.1-foss-2018b
bedtools window -w 10 -a pass.puffin -b fail.puffin -v > pass.puffin.r2
bedtools merge -i pass.puffin.r2 -c 4,6 -o collapse -d 10 | grep -v "," > pass.puffin.r3
bedtools merge -i pass.puffin.r2 -c 4,4,4,6 -o collapse,count_distinct,distinct,collapse -d 10 | grep "," | awk '$5 == 1' | awk -v OFS="\t" '{print $1, $2, $3, $6, $7}' | sed "s/,/\//g" >> pass.puffin.r3

cat pass.puffin.r3 | bedtools sort | awk -v OFS="\t" '{print $1, $2, $3, $4, ($3-$2+1)/$4, "Puffin_STR_"NR, $5}' > Puffin.hipstr_reference.bed

rm fail.puffin filtered_repeats.puffin.bed filtered_repeats.puffin.sorted.bed
rm pass.puffin pass.puffin.r2 pass.puffin.r3

```

Then

```bash

mkdir HipSTR

ls ../Bams/*.Puffin_ReferenceGenome.NU.MT.bam > bamfiles.txt

for chrom in {01..24} Z ;
do echo ${chrom} ;
sbatch HipSTR.sh ${chrom} ;
done
```

HipSTR.sh

```bash


module load HipSTR/0.6.2-foss-2018b-Python-3.6.6

HipSTR --bam-files bamfiles.txt \
--fasta ../Ref/Puffin_ReferenceGenome.NU.MT.FINAL.fasta \
--regions Puffin.hipstr_reference.bed \
--chrom Chr_${1} \
--str-vcf str_calls.Chr_${1}.vcf.gz \
--output-filters

mv str_calls.Chr_${1}.vcf.gz HipSTR/
```

Filtering

```bash

cd HipSTR

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/pyvcf

#filtering according to default HipSTR instruc.
for chrom in {01..24} Z ;
do echo ${chrom} ;
python ../filter_vcf.py --vcf str_calls.Chr_${chrom}.vcf.gz \
--min-call-qual 0.9 \
--max-call-flank-indel 0.15 \
--max-call-stutter 0.15 \
--min-call-allele-bias -2 \
--min-call-strand-bias -2 > str_calls.Chr_${chrom}.filtered.vcf.gz ;
mv str_calls.Chr_${chrom}.vcf.gz str_calls.Chr_${chrom}.vcf ;
gzip str_calls.Chr_${chrom}.vcf ;
mv str_calls.Chr_${chrom}.filtered.vcf.gz str_calls.Chr_${chrom}.filtered.vcf ;
done ;

conda deactivate

#Merging
module purge
module load BCFtools/1.14-GCC-11.2.0

touch samplelist
for chrom in {01..24} Z ;
do echo ${chrom} ;
bgzip str_calls.Chr_${chrom}.filtered.vcf
echo -e "str_calls.Chr_${chrom}.filtered.vcf.gz" >> samplelist ;
done ;

bcftools concat -n -f samplelist -O z -o Puffin_STR_calls.AllChr.filtered.vcf.gz
bcftools index Puffin_STR_calls.AllChr.filtered.vcf.gz
zcat Puffin_STR_calls.AllChr.filtered.vcf.gz | grep -v "#" | wc -l


#filter
bcftools view -i 'INFO/AC>3 && INFO/AN=36' -O z -o Puffin_STR_calls.AllChr.filtered2.vcf.gz Puffin_STR_calls.AllChr.filtered.vcf.gz
zcat Puffin_STR_calls.AllChr.filtered2.vcf.gz | grep -v "#" | wc -l
#covered in all individuals
#maf filter >0.1 (at least 4 alt. alleles)

zcat Puffin_STR_calls.AllChr.filtered2.vcf.gz | grep "#" > header
zcat Puffin_STR_calls.AllChr.filtered2.vcf.gz | grep -v "#" > body

# remove all non STR-length variants
# need to be filtered by checking how many times motif(s) appears in ref and alt
touch body_edit

cat body | while read line
do ID=$(echo ${line} | cut -d ' ' -f3 | sed 's/ //g')
echo $ID
MotifLine=$(grep -w ${ID} ../Puffin.hipstr_reference.bed )
Period=$(echo ${MotifLine} | cut -d ' ' -f4)
if [[ ${Period} -lt 2 ]]
then
	RefSeq=$(echo ${line} | cut -d ' ' -f4)
	Ref_NoMotif_total=0
	while read Motif
		do Ref_NoMotif=$(echo ${RefSeq} | fold -w1 | uniq -c | grep -w "${Motif}" | sed 's/ //g' | awk -F "${Motif}" '{print $1}' | sort | tail -1)
		Ref_NoMotif_total=$(bc <<<"scale=1; $Ref_NoMotif_total + $Ref_NoMotif") 
		done <<< "$(echo ${MotifLine} | cut -d ' ' -f7 | tr '/' '\n' )"
	counter=0
	while read Alt
		do Alt_NoMotif_total=0
		while read Motif ;
			do Alt_NoMotif=$(echo ${Alt} | fold -w1 | uniq -c | grep -w "${Motif}" | sed 's/ //g' | awk -F "${Motif}" '{print $1}' | sort | tail -1)
			Alt_NoMotif_total=$(bc <<<"scale=1; $Alt_NoMotif_total + $Alt_NoMotif")
			done <<< "$(echo ${MotifLine} | cut -d ' ' -f7 | tr '/' '\n' )"
		if [[ ${Alt_NoMotif_total} -ne ${Ref_NoMotif_total} ]]
		then
			counter=$(bc <<<"scale=1; $counter + 1")
		fi
		done <<< "$(echo ${line} | cut -d ' ' -f5 | tr ',' '\n' )"

else
	RefSeq=$(echo ${line} | cut -d ' ' -f4)
	RefSeq_edit=$(echo -e "X${RefSeq}X")
	Ref_NoMotif_total=0
	while read Motif
		do Ref_NoMotif=$(echo $RefSeq_edit | awk -F "$Motif" '{for(i=0;i<=NF;i++){if(!$i){C++}};print C;C=0}')
		if [[ $Ref_NoMotif > 0 ]]
		then
			Ref_NoMotif_edit=$(bc <<<"scale=1; $Ref_NoMotif + 1")
		else
			Ref_NoMotif_edit=1
		fi
		Ref_NoMotif_total=$(bc <<<"scale=1; $Ref_NoMotif_total + $Ref_NoMotif_edit")
		done <<< "$(echo ${MotifLine} | cut -d ' ' -f7 | tr '/' '\n')"
	counter=0
	while read Alt
		do Alt_edit=$(echo -e "X${Alt}X")
		Alt_NoMotif_total=0
		while read Motif
			do Alt_NoMotif=$(echo ${Alt_edit} | awk -F "$Motif" '{for(i=0;i<=NF;i++){if(!$i){C++}};print C;C=0}' | sed 's/ /0/g' )
			if [[ $Alt_NoMotif > 0 ]]
			then
				Alt_NoMotif_edit=$(bc <<<"scale=1; $Alt_NoMotif + 1")
			else
				Alt_NoMotif_edit=1
			fi
			Alt_NoMotif_total=$(bc <<<"scale=1; $Alt_NoMotif_total + $Alt_NoMotif_edit")
			done <<< "$(echo ${MotifLine} | cut -d ' ' -f7 | tr '/' '\n')"
		if [[ ${Alt_NoMotif_total} -ne ${Ref_NoMotif_total} ]]
		then
			counter=$(bc <<<"scale=1; $counter + 1")
		fi
		done <<< "$(echo ${line} | cut -d ' ' -f5 | tr ',' '\n')"
fi
if [[ ${counter} -gt 0 ]]
then
	echo ${line} >> body_edit 
fi
done


cat header body_edit > Puffin_STR_calls.AllChr.filtered3.vcf
rm header
rm body
rm body_edit
gzip Puffin_STR_calls.AllChr.filtered3.vcf
zcat Puffin_STR_calls.AllChr.filtered3.vcf.gz | grep -v "#" | wc -l

```

SMART PCA

```bash

#make Eigenstrat files for smartPCA
zcat Puffin_STR_calls.AllChr.filtered3.vcf.gz | grep "#" > header
zcat Puffin_STR_calls.AllChr.filtered3.vcf.gz | grep -v "#" > body

#geno
cat ../../samples | while read sample ;
do echo $sample ;
column_no=$(tail -1 header | tr '\t' '\n' | nl | grep ${sample} | cut -f1 | sed 's/ //g') ; 
echo -e "${sample} ${sample} 0 0 0 -9" > ${sample}.txt;
cat body | while read line ;
do No_alt_alleles=$(echo ${line} | awk -F ';AC=' '{print $2}' | awk -F ';' '{print $1}' | tr ',' '\n' | wc -l ) ;
No_total_alleles=$(bc <<<"scale=1; $No_alt_alleles + 1") ;
Entry=$(seq 1 ${No_total_alleles} | xargs -I{} echo "0" | tr '\n' ' ') ;
First_allele=$(echo ${line} | awk -F ' ' -v col="$column_no" '{print $col}' | awk -F ':' '{print $1}' | awk -F '|' '{print $1}' ) ;
First_allele_count=$(bc <<<"scale=1; $First_allele + 1" ) ;
Second_allele=$(echo ${line} | awk -F ' ' -v col="$column_no" '{print $col}' | awk -F ':' '{print $1}' | awk -F '|' '{print $2}' ) ;
Second_allele_count=$(bc <<<"scale=1; $Second_allele + 1" ) ;
Entry_1=$(echo ${Entry} | awk -v col="$First_allele_count" -F ' ' '{$col+=1;}1' OFS=' ' ) ;
Entry_2=$(echo ${Entry_1} | awk -v col="$Second_allele_count" -F ' ' '{$col+=1;}1' OFS=' ' ) ;
echo -e "${Entry_2}" | paste -d ' ' ${sample}.txt - > tmp.txt ;
mv tmp.txt ${sample}.txt ;
done;
done ;


touch Puffin_STR_calls.AllChr.geno
cat ../../samples | while read sample ;
do echo $sample ;
cat ${sample}.txt | tr ' ' '\n' | tail -n+7 | paste -d '' Puffin_STR_calls.AllChr.geno - > tmp.txt ;
mv tmp.txt Puffin_STR_calls.AllChr.geno ;
done

touch Puffin_STR_calls.AllChr_NoRelate.geno
cat ../../samples | grep -v "SPI002" | while read sample ;
do echo $sample ;
cat ${sample}.txt | tr ' ' '\n' | tail -n+7 | paste -d '' Puffin_STR_calls.AllChr_NoRelate.geno - > tmp.txt ;
mv tmp.txt Puffin_STR_calls.AllChr_NoRelate.geno ;
done
 

#snp
touch Puffin_STR_calls.AllChr.snp;
cat body | while read line ;
do No_alt_alleles=$(echo ${line} | awk -F ';AC=' '{print $2}' | awk -F ';' '{print $1}' | tr ',' '\n' | wc -l ) ;
No_total_alleles=$(bc <<<"scale=1; $No_alt_alleles + 1") ;
Chr=$(echo ${line} | cut -d ' ' -f1 | sed 's/Chr_//g') ;
ID=$(echo ${line} | cut -d ' ' -f3 | sed 's/ //g') ;
Pos=$(echo ${line} | cut -d ' ' -f2 | sed 's/ //g') ;
seq 1 ${No_total_alleles} | while read number ;
do echo -e "${ID}_${number}\t${Chr}\t0.0\t${Pos}" >> Puffin_STR_calls.AllChr.snp ;
done ;
done ;

#ind
touch Puffin_STR_calls.AllChr.ind
cat ../../samples | while read sample ;
do echo $sample ;
pop=$(echo ${sample} | cut -c1-3) ;
echo -e "${sample}\tU\t${pop}" >> Puffin_STR_calls.AllChr.ind ;
done 

touch Puffin_STR_calls.AllChr_NoRelate.ind
cat ../../samples | grep -v "SPI002" | while read sample ;
do echo $sample ;
pop=$(echo ${sample} | cut -c1-3) ;
echo -e "${sample}\tU\t${pop}" >> Puffin_STR_calls.AllChr_NoRelate.ind ;
done 

#only Autosomes

grep -v -w "Z" Puffin_STR_calls.AllChr.snp > Puffin_STR_calls.Autosomes.snp
wc -l Puffin_STR_calls.Autosomes.snp
head -67230 Puffin_STR_calls.AllChr.geno > Puffin_STR_calls.Autosomes.geno
head -67230 Puffin_STR_calls.AllChr_NoRelate.geno > Puffin_STR_calls.Autosomes_NoRelate.geno

conda activate eigensoft

smartpca -p PuffinSTR_Autosomes_PCA_withoutlier.par > PuffinSTR_Autosomes_PCALog_withoutlier.log
smartpca -p PuffinSTR_Autosomes_PCA_withoutlier_NoRelate.par > PuffinSTR_Autosomes_NoRelate_PCALog_withoutlier.log

conda deactivate

```

Jost D

```bash

gunzip Puffin_STR_calls.AllChr.filtered3.vcf.gz

#only Autosomes and remove SPI002
grep -v -w "Chr_Z" Puffin_STR_calls.AllChr.filtered3.vcf > Puffin_STR_calls.AllChr.filtered4.vcf
grep -v "#" Puffin_STR_calls.AllChr.filtered4.vcf | wc -l 

grep "#" Puffin_STR_calls.AllChr.filtered4.vcf | cut -f1-22,24-27 > header
grep -v "#" Puffin_STR_calls.AllChr.filtered4.vcf | cut -d ' ' -f1-22,24-27 > body

cat body | tr ' ' '\t' > body_new
cat header body_new > Puffin_STR_calls.AllChr.filtered4.edit.vcf

gzip Puffin_STR_calls.AllChr.filtered4.edit.vcf


cat Puffin.hipstr_reference.bed | cut -f4,6 > Puffin.hipstr_reference.bed2
touch Puffin.hipstr_reference.edit.bed2
grep -v "#" Puffin_STR_calls.AllChr.filtered4.vcf | cut -d ' ' -f3 > IDs ;
grep -w -f IDs Puffin.hipstr_reference.bed2 >> Puffin.hipstr_reference.edit.bed2 ;

grep -v "#" Puffin_STR_calls.AllChr.filtered4.vcf | awk -F 'END=' '{print $2}' | awk -F ';' '{print $1}' > Ends

```